pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+  together with their ordering, allowing class patterns to rely on positional
+  sub-patterns without having to explicitly name the attributes in question.
+  This is particularly handy for smaller objects or instances data classes,
+  where the attributes of interest are rather obvious and often have a
+  well-defined ordering.  In a way, ``__match_args__`` is similar to the
+  declaration of formal parameters, which allows to call functions with
+  positional arguments rather than naming all the parameters.
+
+
+The syntax of class patterns is based on the idea that de-construction
+mirrors the syntax of construction.  This is already the case in virtually
+any Python construct, be assignment targets, function definitions or
+iterable unpacking.  In all these cases, we find that the syntax for
+sending and that for receiving 'data' are virtually identical.
+
+- Assignment targets such as variables, attributes and subscripts: 
+  ``foo.bar[2] = foo.bar[3]``;
+
+- Function definitions: a function defined with ``def foo(x, y, z=6)``
+  is called as, e.g., ``foo(123, y=45)``, where the actual arguments
+  provided at the call site are matched against the formal parameters
+  at the definition site;
+
+- Iterable unpacking: ``a, b = b, a`` or ``[a, b] = [b, a]`` or
+  ``(a, b) = (b, a)``, just to name a few equivalent possibilities.
+
+Using the same syntax for reading and writing, l- and r-values, or
+construction and de-construction is widely accepted for its benefits in
+thinking about data, its flow and manipulation.  This equally extends to
+the explicit construction of instances, where class patterns ``c(p, q)``
+deliberately mirror the syntax of creating instances.
+
+
+
+History and Context
+===================
+
+Pattern matching emerged in the late 1970s in the form of tuple unpacking
+and as a means to handle recursive data structures such as linked lists or
+trees (object-oriented languages use the visitor pattern for handling
+recursive data structures).  The early proponents of pattern matching
+organised structured data in 'tagged tuples' rather than ``struct``s as in
+_C_ or the objects introduced later.  A node in a binary tree would, for
+instance, be a tuple with two elements for the left and right branches,
+respectively, and a ``Node``-tag, written as ``Node(left, right)``.  In
+Python we would probably put the tag inside the tuple as
+``('Node', left, right)`` or define a data class `Node` to achieve the
+same effect.
+
+Using modern syntax, a depth-first search (DFS) would then be written as
+follows::
+
+  def DFS(node):
+      node match:
+          case Node(left, right):
+              DFS(left)
+              DFS(right)
+          case Leaf(value):
+              handle(value)
+
+The notion of handling recursive data structures with pattern matching
+immediately gave rise to the idea of handling general recursive patterns
+with pattern matching.  Pattern matching would thus also be used to define
```

Comment:
What I meant is that the base case might not be a specific data class, but could be any value, really.  In other words: the earliest examples I found basically only had constructor and variable patterns, which was certainly enough to handle recursive data structures.  But integers are not exactly recursive data types (or at least that is not the way we think about it).  Moreover, integers tend to form a large "quasi-open" set, as opposed to the data structures with just a handful of different possible forms.

The differene might perhaps be best described through the question: how hard is it to check whether the match statement is exhaustive?

Anyway,  obviously need to reword this.

Reply:
Okay, it's definitely good to know that historically constructor (i.e., class) patterns and variable (capture) patterns are the oldest. If you have the information, I'm curious whether the earliest uses had exhaustiveness checking in mind? That would require either "sealed" classes or global knowledge of all possible classes (which at the scale of programs in the 70s might not have seen as insurmountable :-). I guess tuples also theoretically create a large quasi-open set (though Scala somehow imposes a limit of 22 items...).

Reply:
The question of knowing all possible classes was answered through static typing and because the abstract base class was defined as the union of all possible case classes.  It is, so to speak, the other way round as modern OOP.  

OOP:
```
class Vehicle(): pass
class Car(Vehicle): ...
class Plane(Vehicle): ...
class Train(Vehicle): ...
```
"Functional" style:
```
class Vehicle = Car | Plane | Train
```

And, although it is just a sideline: Scala's limit stems from the fact that tuples need to be defined statically and strongly typed in terms of (generic) classes.  In order to keep meaningful types, you don't want to just use a generic array of objects, but need to allow each item in the tuple to have a different type.  There is basically no way around defining it labourously like this (where the `[T]` introduces a type parameter, written as `<T>` in Java and C):
```
class Tuple2[T1, T2](item1: T1, item2: T2)
class Tuple3[T1, T2, T3](item1: T1, item2: T2, item3: T3)
class Tuple4[T1, T2, T3, T4](item1: T1, item2: T2, item3: T3, item4: T4)
```
At some point you simply have to stop...  And for (unnamed) tuples, I would consider 22 to be a rather generous limit ;-).